### ISSUE TITLE
Compiler Bug regarding Function Hoisting

### ISSUE THREAD

**Issue by TobiasWienand** [googleprojectzero/fuzzilli#488]
Here is a before and after after compiling before.js to after.js
```javascript
// before.js

console.log(two);
function two() {
}
```
```javascript
// after.js

console.log(two);
function f3() {
}
```

**Comment by saelo** [comment]
Right, hoisting doesn't currently work as it's somewhat fundamentally related to variable names, which we loose during compilation. But I think we're pretty close to supporting this: https://github.com/googleprojectzero/fuzzilli/pull/486 should be the first part and I have a local patch set that adds a `.functionName` property to FuzzIL functions. These two taken together should allow us to properly support function hoisting :)

**Comment by TobiasWienand** [comment]
I think this was solved with the most recent commits
